_posts/2023-09-19-virtual-thread-1.adoc

```diff
@@ -0,0 +1,281 @@
+
+Unfortunately, as of today, lots of Java libraries are pinning the carrier thread. 
+The Quarkus team and Red Hat, in general, have patched many libraries (such as Narayana (the transaction manager of Quarkus) or Hibernate ORM) to avoid pinning. 
+However, when you use a library, be careful. 
+It will take time until all the code gets reworked in a more virtual-thread-friendly way.
+
+### 2. Monopolisation
+
+As for the reactive model, if the virtual thread executes intensive and long computation, it monopolizes that carrier.
+The virtual thread scheduler is not preemptive.
+So it cannot interrupt a running thread. 
+It needs to wait for an I/O or the completion of the computation.
+Until then, this carrier thread cannot execute other virtual threads.
+Using a dedicated platform thread pool might be wiser when executing long computations. 
+
+### 3. Carrier thread pool elasticity
+
+When there is pinning or monopolization, the JVM may create new carrier threads to avoid having too many unscheduled virtual threads. 
+
+These creations are creating platform/OS threads. 
+So, it's expensive and uses memory.
+You especially need to pay attention to the second point. You may hit the memory limit if you run on low resources and your code is not very virtual-thread-friendly. 
+
+In a container with memory constraints, the application can be killed.
+So, always check for pinning, monopolization, and memory usage.
+
+### 4. Object pooling
+
+For years, threads were scarce resources.
+It was recommended to pool them and reuse them. 
+This good practice has encouraged the use of thread locals as an object-pooling mechanism.
+Like Jackson or Netty, many libraries store expensive objects in thread locals.
+These objects can only be accessed by the code running on the thread in which the objects are stored.
+Because the number of threads was limited, it capped the number of creation.
+Also, because threads were reused, the objects were cached and reused. 
+Unfortunately, these two assumptions are not valid with virtual threads:
+You can have a lot of them
+There are not reused
+Thus, libraries utilizing these pooling patterns may underperform when using virtual threads. 
+You will see many allocations of large objects, as every virtual thread will get its own instance of the object. 
+
+### 5. Stressing thread safety
+
+Virtual threads provide a new way to build concurrent applications in Java.
+You are not limited by the number of threads in the pool.
+You do not have to use asynchronous development models. 
+
+But, before rewriting your application to leverage this new mechanism, ensure the code is thread-safe. 
+Many libraries and frameworks do not allow concurrent access to some objects.
+For example, database connections should not be accessed concurrently. 
+You must be cautious when you have many virtual threads, especially when using the structured concurrency API (still in preview in Java 21).
```

Comment:
It would be nice to have some more juice on why structured concurrency might be a challenge.

Reply:
I've added an example.